### ISSUE TITLE
IfcOpenShell creates entities with STEP ID 0 after adding to a file from another

### ISSUE THREAD

**Issue by Moult** [IfcOpenShell/IfcOpenShell#999]
I have `fileA`, which is a new file I'm creating, and `fileB`, which contains some IFC entities, but not a full project (i.e. no `IfcProject`).

So I'm creating entities in `fileA` with `fileA.createIfcFoo(...)` and `fileA.create_entity(...)`, without issues.

Then, halfway through that process, I try to do `fileA.add(x)` where `x` comes from `fileB`. This add process seems to also finish without issues.

However, it seems as though once I've `add`ed some of these element, then when I try to resume creating entities in `fileA` with `fileA.createIfcFoo(...)` and `fileA.create_entity(...)` again, I get results like this:

```
#0=IfcSurfaceStyleRendering(#35,-0.,$,$,$,$,$,$,.NOTDEFINED.)
#0=IfcSurfaceStyle('Material.001',.BOTH.,(#0))
#0=IfcStyledItem($,(#0),'Material.001')
```

Sorry for the lack of a test case, I haven't been able to create a minimum example. It also seems as though the number of zero STEP IDs I get corresponds to how many entities I've added.

Any ideas why this type of behaviour might happen?

**Comment by aothms** [comment]
I think the instances start with 0 and get assigned a id when added to the file according to https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcparse/IfcFile.h#L193 . So apparently it's the file.add() part that's failing and they keep their original id set to zero. Looking forward to more info.

**Comment by Moult** [comment]
@aothms I created a test case, and I have NO idea what is going on :)

First, this is a test case which works as expected:

```
class Foo():
    def __init__(self):
        self.file = ifcopenshell.file()

    def start(self):
        self.foo()

        dummy = ifcopenshell.file()
        for i in range(0, 20):
            dummy.createIfcWall()
        for element in dummy:
            self.file.add(element)

        self.foo()

    def foo(self):
        for i in range(0, 60):
            rgb = self.file.createIfcSlab()
            print(rgb)

foo = Foo()
foo.start()
```

If I run that, I get this:

```
#1=IfcSlab($,$,$,$,$,$,$,$,$)
#2=IfcSlab($,$,$,$,$,$,$,$,$)
#3=IfcSlab($,$,$,$,$,$,$,$,$)
... snip, because it's boring ...
#138=IfcSlab($,$,$,$,$,$,$,$,$)
#139=IfcSlab($,$,$,$,$,$,$,$,$)
#140=IfcSlab($,$,$,$,$,$,$,$,$)
```

Now, here's a variation of the code, where all I do is move the `add()` code into it's own function:

```
class Foo():
    def __init__(self):
        self.file = ifcopenshell.file()

    def start(self):
        self.foo()
        self.bar()
        self.foo()

    def foo(self):
        for i in range(0, 60):
            rgb = self.file.createIfcSlab()
            print(rgb)

    def bar(self):
        dummy = ifcopenshell.file()
        for i in range(0, 20):
            dummy.createIfcWall()
        for element in dummy:
            self.file.add(element)

foo = Foo()
foo.start()
```

Now I get this:

```
#1=IfcSlab($,$,$,$,$,$,$,$,$)
#2=IfcSlab($,$,$,$,$,$,$,$,$)
#3=IfcSlab($,$,$,$,$,$,$,$,$)
... snip ... it works well until ID 60 ... then ...
#58=IfcSlab($,$,$,$,$,$,$,$,$)
#59=IfcSlab($,$,$,$,$,$,$,$,$)
#60=IfcSlab($,$,$,$,$,$,$,$,$)
#0=IfcSlab($,$,$,$,$,$,$,$,$)
#81=IfcSlab($,$,$,$,$,$,$,$,$)
#82=IfcSlab($,$,$,$,$,$,$,$,$)
#83=IfcSlab($,$,$,$,$,$,$,$,$)
#84=IfcSlab($,$,$,$,$,$,$,$,$)
#85=IfcSlab($,$,$,$,$,$,$,$,$)
#0=IfcSlab($,$,$,$,$,$,$,$,$)
#86=IfcSlab($,$,$,$,$,$,$,$,$)
#87=IfcSlab($,$,$,$,$,$,$,$,$)
#88=IfcSlab($,$,$,$,$,$,$,$,$)
#89=IfcSlab($,$,$,$,$,$,$,$,$)
#90=IfcSlab($,$,$,$,$,$,$,$,$)
#91=IfcSlab($,$,$,$,$,$,$,$,$)
#92=IfcSlab($,$,$,$,$,$,$,$,$)
#93=IfcSlab($,$,$,$,$,$,$,$,$)
#0=IfcSlab($,$,$,$,$,$,$,$,$)
#0=IfcSlab($,$,$,$,$,$,$,$,$)
#0=IfcSlab($,$,$,$,$,$,$,$,$)
#94=IfcSlab($,$,$,$,$,$,$,$,$)
#95=IfcSlab($,$,$,$,$,$,$,$,$)
#0=IfcSlab($,$,$,$,$,$,$,$,$)
#0=IfcSlab($,$,$,$,$,$,$,$,$)
#96=IfcSlab($,$,$,$,$,$,$,$,$)
#97=IfcSlab($,$,$,$,$,$,$,$,$)
#0=IfcSlab($,$,$,$,$,$,$,$,$)
#98=IfcSlab($,$,$,$,$,$,$,$,$)
#0=IfcSlab($,$,$,$,$,$,$,$,$)
#99=IfcSlab($,$,$,$,$,$,$,$,$)
#100=IfcSlab($,$,$,$,$,$,$,$,$)
... notice in the above there are some zeros... then it continues working until the end ...
#129=IfcSlab($,$,$,$,$,$,$,$,$)
#130=IfcSlab($,$,$,$,$,$,$,$,$)
#131=IfcSlab($,$,$,$,$,$,$,$,$)
```

**Comment by Moult** [comment]
This is using the latest IfcOpenBot build on Linux, and I was also able to recreate this on a Windows machine. The exact creation of the zeros are different on different machines, but it seems to always occur after the `add()` statements occur. It might be my imagination but I think the number of zeros also correlates with the number of `add()`s.

**Comment by Moult** [comment]
@aothms  you're a genius! Thanks so much!

**Comment by Moult** [comment]
Elaborating further on this, here's another snippet where `bar()` is called twice. The second time `bar()` is called, `self.dummy` is garbage collected and funky numbers come out - not zeros, but funky ones.

```
class Foo():
    def __init__(self):
        self.file = ifcopenshell.file()

    def start(self):
        self.foo()
        self.bar()
        self.bar()
        self.foo()

    def foo(self):
        for i in range(0, 60):
            rgb = self.file.createIfcSlab()
            print(rgb)

    def bar(self):
        print('bar')
        self.dummy = ifcopenshell.file()
        for i in range(0, 20):
            self.dummy.createIfcColumn()
        for element in self.dummy:
            x = self.file.add(element)
            print(x)


foo = Foo()
foo.start()
```

Results:

```
#1=IfcSlab($,$,$,$,$,$,$,$,$)
#2=IfcSlab($,$,$,$,$,$,$,$,$)
#3=IfcSlab($,$,$,$,$,$,$,$,$)
... snip ... happy until ...
#58=IfcSlab($,$,$,$,$,$,$,$,$)
#59=IfcSlab($,$,$,$,$,$,$,$,$)
#60=IfcSlab($,$,$,$,$,$,$,$,$)
bar
#61=IfcColumn($,$,$,$,$,$,$,$,$)
#62=IfcColumn($,$,$,$,$,$,$,$,$)
#63=IfcColumn($,$,$,$,$,$,$,$,$)
#64=IfcColumn($,$,$,$,$,$,$,$,$)
#65=IfcColumn($,$,$,$,$,$,$,$,$)
#66=IfcColumn($,$,$,$,$,$,$,$,$)
#67=IfcColumn($,$,$,$,$,$,$,$,$)
#68=IfcColumn($,$,$,$,$,$,$,$,$)
#69=IfcColumn($,$,$,$,$,$,$,$,$)
#70=IfcColumn($,$,$,$,$,$,$,$,$)
#71=IfcColumn($,$,$,$,$,$,$,$,$)
#72=IfcColumn($,$,$,$,$,$,$,$,$)
#73=IfcColumn($,$,$,$,$,$,$,$,$)
#74=IfcColumn($,$,$,$,$,$,$,$,$)
#75=IfcColumn($,$,$,$,$,$,$,$,$)
#76=IfcColumn($,$,$,$,$,$,$,$,$)
#77=IfcColumn($,$,$,$,$,$,$,$,$)
#78=IfcColumn($,$,$,$,$,$,$,$,$)
#79=IfcColumn($,$,$,$,$,$,$,$,$)
#80=IfcColumn($,$,$,$,$,$,$,$,$)
bar
#65=IfcColumn($,$,$,$,$,$,$,$,$)
#71=IfcColumn($,$,$,$,$,$,$,$,$)
#68=IfcColumn($,$,$,$,$,$,$,$,$)
#80=IfcColumn($,$,$,$,$,$,$,$,$)
#81=IfcColumn($,$,$,$,$,$,$,$,$)
#67=IfcColumn($,$,$,$,$,$,$,$,$)
#82=IfcColumn($,$,$,$,$,$,$,$,$)
#66=IfcColumn($,$,$,$,$,$,$,$,$)
#70=IfcColumn($,$,$,$,$,$,$,$,$)
#75=IfcColumn($,$,$,$,$,$,$,$,$)
#69=IfcColumn($,$,$,$,$,$,$,$,$)
#73=IfcColumn($,$,$,$,$,$,$,$,$)
#83=IfcColumn($,$,$,$,$,$,$,$,$)
#72=IfcColumn($,$,$,$,$,$,$,$,$)
#78=IfcColumn($,$,$,$,$,$,$,$,$)
#84=IfcColumn($,$,$,$,$,$,$,$,$)
#77=IfcColumn($,$,$,$,$,$,$,$,$)
#76=IfcColumn($,$,$,$,$,$,$,$,$)
#79=IfcColumn($,$,$,$,$,$,$,$,$)
#64=IfcColumn($,$,$,$,$,$,$,$,$)
#85=IfcSlab($,$,$,$,$,$,$,$,$)
#86=IfcSlab($,$,$,$,$,$,$,$,$)
#87=IfcSlab($,$,$,$,$,$,$,$,$)
... snip ...
#142=IfcSlab($,$,$,$,$,$,$,$,$)
#143=IfcSlab($,$,$,$,$,$,$,$,$)
#144=IfcSlab($,$,$,$,$,$,$,$,$)
```

**Comment by Moult** [comment]
Is it still relevant to keep this open, or should it be closed?

**Comment by aothms** [comment]
> The exact behaviour I find hard to explain though, dummy should be allowed to be garbage collected. Apparently on a lower level some data is still shared between the instances in dummy and the copied ones in self.file. I find it a bit puzzling how it affects the numeric identifiers.

Apparently at that point I didn't see the issue. I'm quite sure we addressed this in #1732

> Is it still relevant to keep this open, or should it be closed?

I think it can be closed.
